README.md
# Paranoidhttp

[![Build Status](https://travis-ci.org/hakobe/paranoidhttp.svg?branch=master)][travis]
[![Coverage Status](https://coveralls.io/repos/hakobe/paranoidhttp/badge.svg?branch=master)][coveralls]
[![MIT License](http://img.shields.io/badge/license-MIT-blue.svg?style=flat-square)][license]
[![GoDoc](https://godoc.org/github.com/hakobe/paranoidhttp?status.svg)][godoc]

[travis]: https://travis-ci.org/hakobe/paranoidhttp
[coveralls]: https://coveralls.io/r/hakobe/paranoidhttp?branch=master
[license]: https://github.com/hakobe/paranoidhttp/blob/master/LICENSE
[godoc]: https://godoc.org/github.com/hakobe/paranoidhttp

Paranoidhttp provides a pre-configured http.Client that protects you from harm.

## Description

Paranoidhttp is a factory of http.Client that is paranoid againt attackers.
This is useful when you create an HTTP request using inputs from possibly malicious users.

The created http.Client protects you from connecting to internal IP ranges
even though redirects or DNS tricks are used.

## Synopsis

```go
// use the default client for ease
res, err := paranoidhttp.DefaultClient.Get("http://www.hatena.ne.jp")

// or customize the client for yourself
client, transport, dialer := paranoidhttp.NewClient()
client.Timeout = 10 * time.Second
transport.DisableCompression = true
dialer.KeepAlive = 60 * time.Second

// Add an permitted ipnets with functional option
ipNet, _ := net.ParseCIDR("127.0.0.1/32")
client, _, _ := paranoidhttp.New(
    paranoidhttp.PermittedIPNets(ipNet))
```

## Acknowledgement

I want to thank [LWPx::ParanoidAgent](https://metacpan.org/pod/LWPx::ParanoidAgent).

## License

[MIT](./LICENSE)

## Author

[hakobe](http://github.com/hakobe)


CHANGELOG.md
# Changelog

## [v0.3.0](https://github.com/hakobe/paranoidhttp/compare/v0.2.0...v0.3.0) (2023-01-19)

* Support ipv6 in README [#16](https://github.com/hakobe/paranoidhttp/pull/16) ([karmanyaahm](https://github.com/karmanyaahm))
* Add IPv6 support [#15](https://github.com/hakobe/paranoidhttp/pull/15) ([karmanyaahm](https://github.com/karmanyaahm))
* Fix IPv6 unspecified address [#12](https://github.com/hakobe/paranoidhttp/pull/12) ([mvisat](https://github.com/mvisat))

## [v0.2.0](https://github.com/hakobe/paranoidhttp/compare/v0.1.0...v0.2.0) (2019-04-02)

* Make the list of cidrs and hosts configurable [#10](https://github.com/hakobe/paranoidhttp/pull/10) ([Songmu](https://github.com/Songmu))

## [v0.1.0](https://github.com/hakobe/paranoidhttp/compare/6495e8792661...v0.1.0) (2019-03-31)

* adjust releng files [#11](https://github.com/hakobe/paranoidhttp/pull/11) ([Songmu](https://github.com/Songmu))
* Revert "introduce structure for errors" [#8](https://github.com/hakobe/paranoidhttp/pull/8) ([hakobe](https://github.com/hakobe))
* introduce structure for errors [#7](https://github.com/hakobe/paranoidhttp/pull/7) ([astj](https://github.com/astj))
* Remove resolved issues from readme. [#6](https://github.com/hakobe/paranoidhttp/pull/6) ([shogo82148](https://github.com/shogo82148))
* fix errors in IPv6 environment. [#5](https://github.com/hakobe/paranoidhttp/pull/5) ([shogo82148](https://github.com/shogo82148))
* [incompatible] Support context more on Go 1.8 [#4](https://github.com/hakobe/paranoidhttp/pull/4) ([Songmu](https://github.com/Songmu))
* support DialContext for go 1.7 [#3](https://github.com/hakobe/paranoidhttp/pull/3) ([Songmu](https://github.com/Songmu))
* define mustParseCIDR and refactor [#2](https://github.com/hakobe/paranoidhttp/pull/2) ([Songmu](https://github.com/Songmu))
* To prevent TOCTOU attacks, use checked IP addr directly for connections [#1](https://github.com/hakobe/paranoidhttp/pull/1) ([hakobe](https://github.com/hakobe))


client_test.go
package paranoidhttp

import (
	"net"
	"testing"
)

func TestRequest(t *testing.T) {
	resp, err := DefaultClient.Get("http://www.example.org")
	if err != nil && resp.StatusCode == 200 {
		t.Error("The request with an ordinal url should be successful")
	}

	resp, err = DefaultClient.Get("http://localhost")
	if err == nil {
		t.Errorf("The request for localhost should be fail")
	}

	if _, err := DefaultClient.Get("http://192.168.0.1"); err == nil {
		t.Errorf("The request for localhost should be fail")
	}

	if _, err := DefaultClient.Get("http://[::]"); err == nil {
		t.Errorf("The request for IPv6 unspecified address should be fail")
	}
	if _, err := DefaultClient.Get("http://[fd00::1234]"); err == nil {
		t.Errorf("The request for IPv6 ULA should fail")
	}
}

func TestIsHostForbidden(t *testing.T) {
	badHosts := []string{
		"localhost",
		"host has space",
	}

	for _, h := range badHosts {
		if !basicConfig().isHostForbidden(h) {
			t.Errorf("%s should be forbidden", h)
		}
	}

	notBadHosts := []string{
		"www.hatena.ne.jp",
		"www.google.com",
		"xn--t8jx73hngb.jp",
	}

	for _, h := range notBadHosts {
		if basicConfig().isHostForbidden(h) {
			t.Errorf("%s should not be forbidden", h)
		}
	}
}

func TestIsIpForbidden(t *testing.T) {
	badIPs := []string{
		"0.0.0.0", "::", // Unspecified
		"127.0.0.0", "127.255.255.255", "::1", // Loopback
		"10.0.0.0", "10.255.255.255", // Private A
		"172.16.0.0", "172.31.255.255", // Private B
		"192.168.0.0", "192.168.255.255", // Private C
		"fc00::", "fdff:ffff:ffff:ffff:ffff:ffff:ffff:ffff", // Private v6
		"192.0.2.0", "192.0.2.255", // Test-Net
		"192.88.99.0", "192.88.99.255", // 6to4 relay
		"224.0.0.0", "239.255.255.255", // Multicast
		"ff00::", "ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff", //Multicast 6
		"169.254.0.0", "169.254.255.255", // Link local
		"fe80::", "febf:ffff:ffff:ffff:ffff:ffff:ffff:ffff", // Link Local v6
		"::ffff:0:255.255.255.255", "::ffff:255.255.255.255", //v4 to v6 mapping
		"2001:db8::", "2001:db8:ffff:ffff:ffff:ffff:ffff:ffff", //v6 documentation
	}

	for _, ip := range badIPs {
		if !basicConfig().isIPForbidden(net.ParseIP(ip)) {
			t.Errorf("%s should be forbidden", ip)
		}
	}

	notBadIPs := []string{
		"0.0.0.1", "8.8.8.8",
		"126.255.255.255", "128.0.0.0",
		"9.255.255.255", "11.0.0.0",
		"172.15.255.255", "172.32.0.0",
		"192.167.255.255", "192.169.0.0",
		"192.88.98.255", "192.88.100.0",
		"223.255.255.255", "240.0.0.0",
		"169.253.255.255", "169.255.0.0",
		"2000::1", "3fff::1",
		// real examples
		"2606:4700:4700::1111", "2001:4860:4860::8888",
	}

	for _, ip := range notBadIPs {
		if basicConfig().isIPForbidden(net.ParseIP(ip)) {
			t.Errorf("%s should not be forbidden", ip)
		}
	}

	c := basicConfig()
	ip := "172.18.0.1"
	if !c.isIPForbidden(net.ParseIP(ip)) {
		t.Errorf("%s should be forbidden", ip)
	}

	c.PermittedIPNets = append(c.PermittedIPNets, mustParseCIDR("172.18.0.1/32"))
	if c.isIPForbidden(net.ParseIP(ip)) {
		t.Errorf("%s should not be forbidden", ip)
	}
}


client.go
package paranoidhttp

import (
	"context"
	"errors"
	"fmt"
	"net"
	"net/http"
	"regexp"
	"time"
)

// Config stores the rules for allowing IP/hosts
type config struct {
	ForbiddenIPNets []*net.IPNet
	PermittedIPNets []*net.IPNet
	ForbiddenHosts  []*regexp.Regexp
}

// DefaultClient is the default Client whose setting is the same as http.DefaultClient.
var (
	defaultConfig config
	DefaultClient *http.Client
)

func mustParseCIDR(addr string) *net.IPNet {
	_, ipnet, err := net.ParseCIDR(addr)
	if err != nil {
		panic(`net: ParseCIDR("` + addr + `"): ` + err.Error())
	}
	return ipnet
}

func init() {
	forbiddenIPs := []string{
		//IPv4
		"10.0.0.0/8",     // private class A
		"172.16.0.0/12",  // private class B
		"192.168.0.0/16", // private class C
		"192.0.2.0/24",   // test net 1
		"192.88.99.0/24", // 6to4 relay

		// ipv6
		// block everything except 2000::/3 according to rfc2373#section-2.4
		"0000::/3", // 0000-0010
		"4000::/2", // 0100-1000
		"8000::/1", // 1000-1111
		//v6 special ranges inside 2000::/3
		"2001::/32",     // Teredo tunneling
		"2001:10::/28",  // Deprecated (previously ORCHID)
		"2001:20::/28",  // ORCHIDv2
		"2001:db8::/32", // Addresses used in documentation and example source code
		"2002::/16",     // 6to4
	}

	defaultConfig = config{
		ForbiddenIPNets: make([]*net.IPNet, len(forbiddenIPs)),
		ForbiddenHosts: []*regexp.Regexp{
			regexp.MustCompile(`(?i)^localhost$`),
			regexp.MustCompile(`(?i)\s+`),
		},
	}
	for n, i := range forbiddenIPs {
		defaultConfig.ForbiddenIPNets[n] = mustParseCIDR(i)
	}

	DefaultClient, _, _ = NewClient()
}

// isHostForbidden checks whether a hostname is forbidden by the Config
func (c *config) isHostForbidden(host string) bool {
	for _, forbiddenHost := range c.ForbiddenHosts {
		if forbiddenHost.MatchString(host) {
			return true
		}
	}
	return false
}

// isIPForbidden checks whether an IP address is forbidden by the Config
func (c *config) isIPForbidden(ip net.IP) bool {
	for _, permittedIPNet := range c.PermittedIPNets {
		if permittedIPNet.Contains(ip) {
			return false
		}
	}

	if !ip.IsGlobalUnicast() {
		return true
	}

	for _, forbiddenIPNet := range c.ForbiddenIPNets {
		if forbiddenIPNet.Contains(ip) {
			return true
		}
	}
	return false
}

// BasicConfig contains the most common hosts and IPs to be blocked
func basicConfig() *config {
	c := defaultConfig // copy to return clone
	return &c
}

// Option type of paranoidhttp
type Option func(*config)

// ForbiddenIPNets sets forbidden IPNets
func ForbiddenIPNets(ips ...*net.IPNet) Option {
	return func(c *config) {
		c.ForbiddenIPNets = ips
	}
}

// PermittedIPNets sets permitted IPNets
// It takes priority over other forbidden rules.
func PermittedIPNets(ips ...*net.IPNet) Option {
	return func(c *config) {
		c.PermittedIPNets = ips
	}
}

// ForbiddenHosts set forbidden host rules by regexp
func ForbiddenHosts(hostRegs ...*regexp.Regexp) Option {
	return func(c *config) {
		c.ForbiddenHosts = hostRegs
	}
}

func safeAddr(ctx context.Context, resolver *net.Resolver, hostport string, opts ...Option) (string, error) {
	c := basicConfig()
	for _, opt := range opts {
		opt(c)
	}
	host, port, err := net.SplitHostPort(hostport)
	if err != nil {
		return "", err
	}

	ip := net.ParseIP(host)
	if ip != nil {
		if ip.IsUnspecified() || c.isIPForbidden(ip) {
			return "", fmt.Errorf("bad ip is detected: %v", ip)
		}
		return net.JoinHostPort(ip.String(), port), nil
	}

	if c.isHostForbidden(host) {
		return "", fmt.Errorf("bad host is detected: %v", host)
	}

	r := resolver
	if r == nil {
		r = net.DefaultResolver
	}
	addrs, err := r.LookupIPAddr(ctx, host)
	if err != nil || len(addrs) <= 0 {
		return "", err
	}
	safeAddrs := make([]net.IPAddr, 0, len(addrs))
	for _, addr := range addrs {
		if c.isIPForbidden(addr.IP) {
			return "", fmt.Errorf("bad ip is detected: %v", addr.IP)
		}
		safeAddrs = append(safeAddrs, addr)
	}
	if len(safeAddrs) == 0 {
		return "", fmt.Errorf("fail to lookup ip addr: %v", host)
	}
	return net.JoinHostPort(safeAddrs[0].IP.String(), port), nil
}

// NewDialer returns a dialer function which only accepts connections to secure hosts.
//
// This is used to create a new paranoid http.Client,
func NewDialer(dialer *net.Dialer, opts ...Option) func(ctx context.Context, network, addr string) (net.Conn, error) {
	return func(ctx context.Context, network, hostport string) (net.Conn, error) {
		switch network {
		case "tcp", "tcp4", "tcp6":
			addr, err := safeAddr(ctx, dialer.Resolver, hostport, opts...)
			if err != nil {
				return nil, err
			}
			return dialer.DialContext(ctx, network, addr)
		default:
			return nil, errors.New("does not support any networks except tcp")
		}
	}
}

// NewClient returns a new http.Client configured to be paranoid for attackers.
//
// This also returns http.Tranport and net.Dialer so that you can customize those behavior.
func NewClient(opts ...Option) (*http.Client, *http.Transport, *net.Dialer) {
	dialer := &net.Dialer{
		Timeout:   30 * time.Second,
		KeepAlive: 30 * time.Second,
	}
	transport := &http.Transport{
		Proxy:               http.ProxyFromEnvironment,
		DialContext:         NewDialer(dialer, opts...),
		TLSHandshakeTimeout: 10 * time.Second,
	}
	return &http.Client{
		Timeout:   30 * time.Second,
		Transport: transport,
	}, transport, dialer
}
